cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java

```diff
@@ -1982,22 +2015,42 @@ public ConfigMap generateSharedConfigurationConfigMap(MetricsAndLogging metricsA
-                .withRackId(rack)
-                .withZookeeper(cluster)
-                .withLogDirs(VolumeUtils.createVolumeMounts(storage, mountPath, false))
-                .withListeners(cluster,
-                        namespace,
-                        listeners,
-                        () -> getPodName(brokerId),
-                        listenerId -> advertisedHostnames.get(brokerId).get(listenerId),
-                        listenerId -> advertisedPorts.get(brokerId).get(listenerId),
-                        controlPlaneListener)
-                .withAuthorization(cluster, authorization)
-                .withCruiseControl(cluster, cruiseControlSpec, ccNumPartitions, ccReplicationFactor, ccMinInSyncReplicas)
-                .withUserConfiguration(configuration)
-                .build().trim();
+        if (useKRaft) {
+            return new KafkaBrokerConfigurationBuilder(reconciliation)
+                    .withBrokerId(String.valueOf(brokerId))
+                    .withRackId(rack)
+                    .withKRaft(cluster, namespace, replicas)
+                    .withLogDirs(VolumeUtils.createVolumeMounts(storage, mountPath, false))
+                    .withListeners(cluster,
+                            namespace,
+                            listeners,
+                            () -> getPodName(brokerId),
+                            listenerId -> advertisedHostnames.get(brokerId).get(listenerId),
+                            listenerId -> advertisedPorts.get(brokerId).get(listenerId),
+                            controlPlaneListener, true)
+                    // Not supported right now
+                    //.withAuthorization(cluster, authorization)
+                    .withCruiseControl(cluster, cruiseControlSpec, ccNumPartitions, ccReplicationFactor, ccMinInSyncReplicas)
+                    .withUserConfiguration(configuration)
+                    .build().trim();
+        } else {
+            return new KafkaBrokerConfigurationBuilder(reconciliation)
+                    .withBrokerId(String.valueOf(brokerId))
+                    .withRackId(rack)
+                    .withZookeeper(cluster)
+                    .withLogDirs(VolumeUtils.createVolumeMounts(storage, mountPath, false))
+                    .withListeners(cluster,
+                            namespace,
+                            listeners,
+                            () -> getPodName(brokerId),
+                            listenerId -> advertisedHostnames.get(brokerId).get(listenerId),
+                            listenerId -> advertisedPorts.get(brokerId).get(listenerId),
+                            controlPlaneListener, false)
+                    .withAuthorization(cluster, authorization)
+                    .withCruiseControl(cluster, cruiseControlSpec, ccNumPartitions, ccReplicationFactor, ccMinInSyncReplicas)
+                    .withUserConfiguration(configuration)
+                    .build().trim();
```

Comment:
instead of code duplication can't you instantiate the `KafkaBrokerConfigurationBuilder`, calling all the common `withXXXX` then  applying the `if` on KRaft only for specific settings and finally returning the config?
Something like this for example.

```java
KafkaBrokerConfigurationBuilder builder =
KafkaBrokerConfigurationBuilder(reconciliation)
                    .withBrokerId(String.valueOf(brokerId))
                    .withRackId(rack)
                    .withLogDirs(VolumeUtils.createVolumeMounts(storage, mountPath, false))
                    .withListeners(cluster,
                            namespace,
                            listeners,
                            () -> getPodName(brokerId),
                            listenerId -> advertisedHostnames.get(brokerId).get(listenerId),
                            listenerId -> advertisedPorts.get(brokerId).get(listenerId),
                            controlPlaneListener, false)
                    .withAuthorization(cluster, authorization)
                    .withCruiseControl(cluster, cruiseControlSpec, ccNumPartitions, ccReplicationFactor, ccMinInSyncReplicas)
                    .withUserConfiguration(configuration);

if (useKRaft) {
   builder.withKRaft(cluster, namespace, replicas);
} else {
   builder.withZookeeper(cluster).withAuthorization(cluster, authorization);
}

return builder.build().trim();
```

Anyway it's not mandatory to change if you don't want to do so.

Reply:
TBH, it seemed more _future-proof_ to me. I think they will diverge only more and more in the future with authorization, with the actual KRaft roles being different, etc. So I would probably prefer to keep it the way it is.